usr/src/man/man1/pwd.1
.\"
.\" Sun Microsystems, Inc. gratefully acknowledges The Open Group for
.\" permission to reproduce portions of its copyrighted documentation.
.\" Original documentation from The Open Group can be obtained online at
.\" http://www.opengroup.org/bookstore/.
.\"
.\" The Institute of Electrical and Electronics Engineers and The Open
.\" Group, have given us permission to reprint portions of their
.\" documentation.
.\"
.\" In the following statement, the phrase ``this text'' refers to portions
.\" of the system documentation.
.\"
.\" Portions of this text are reprinted and reproduced in electronic form
.\" in the SunOS Reference Manual, from IEEE Std 1003.1, 2004 Edition,
.\" Standard for Information Technology -- Portable Operating System
.\" Interface (POSIX), The Open Group Base Specifications Issue 6,
.\" Copyright (C) 2001-2004 by the Institute of Electrical and Electronics
.\" Engineers, Inc and The Open Group.  In the event of any discrepancy
.\" between these versions and the original IEEE and The Open Group
.\" Standard, the original IEEE and The Open Group Standard is the referee
.\" document.  The original Standard can be obtained online at
.\" http://www.opengroup.org/unix/online.html.
.\"
.\" This notice shall appear on any product containing this material.
.\"
.\" The contents of this file are subject to the terms of the
.\" Common Development and Distribution License (the "License").
.\" You may not use this file except in compliance with the License.
.\"
.\" You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
.\" or http://www.opensolaris.org/os/licensing.
.\" See the License for the specific language governing permissions
.\" and limitations under the License.
.\"
.\" When distributing Covered Code, include this CDDL HEADER in each
.\" file and include the License file at usr/src/OPENSOLARIS.LICENSE.
.\" If applicable, add the following below this CDDL HEADER, with the
.\" fields enclosed by brackets "[]" replaced with your own identifying
.\" information: Portions Copyright [yyyy] [name of copyright owner]
.\"
.\"
.\" Copyright 1989 AT&T
.\" Portions Copyright (c) 1992, X/Open Company Limited  All Rights Reserved
.\" Portions Copyright (c) 1982-2007 AT&T Knowledge Ventures
.\" Copyright (c) 2007, Sun Microsystems, Inc.  All Rights Reserved
.\" Copyright 2026 Oxide Computer Company
.\"
.Dd March 31, 2026
.Dt PWD 1
.Os
.Sh NAME
.Nm pwd
.Nd return working directory name
.Sh SYNOPSIS
.Nm /usr/bin/pwd
.Op Fl L | P
.Sh DESCRIPTION
The
.Nm
utility writes an absolute path name of the current working directory to
standard output.
.Pp
The Bourne shell,
.Xr sh 1 ,
the Korn shells,
.Xr ksh 1
and
.Xr ksh93 1 ,
and most other shells have a built-in
.Ic pwd
command.
If just
.Ic pwd
is used while in shell context and not an absolute path to a command, the
shell's built-in version will be executed.
.Pp
When no options are specified,
.Nm
operates as though
.Fl P
was specified.
Systems vary in the defaults they use here and many historical defaults diverge
from POSIX.
For portable operation, always specify
.Fl L
or
.Fl P .
.Sh OPTIONS
.Nm pwd
supports the following options:
.Bl -tag -width Fl
.It Fl L
Output a logical path.
The
.Ev PWD
environment variable will be used as long as it appears valid.
Criteria may include whether this is an absolute or relative path, if there are
any path component's that are
.Sq \&. .
or
.Sq .. ,
and if the actual path refers to the expected file system location.
.Pp
The
.Fl L
option overrides any prior instances of the
.Fl P
option.
.It Fl P
Print the physical path, resolves any symlinks encountered.
The
.Fl P
option overrides any prior instances of the
.Fl L
option.
.El
.Sh ENVIRONMENT VARIABLES
.Bl -tag -width Ev
.It Ev PWD
The logical current working directory which is used when the
.Fl L
option is specified.
.El
.Pp
In addition, see
.Xr environ 7
for descriptions of the following environment variables that affect the
execution of
.Nm :
.Ev LANG ,
.Ev LC_ALL ,
.Ev LC_MESSAGES ,
and
.Ev NLSPATH .
.Sh EXIT STATUS
The following exit values are returned:
.Bl -tag -width Ds
.It Sy 0
Successful completion.
.It Sy >0
An error occurred.
.El
.Pp
If an error is detected, output will not be written to standard output, a
diagnostic message will be written to standard error, and the exit status will
not be
.Sy 0 .
.Sh DIAGNOSTICS
.Dq Sy Cannot open ..
and
.Dq Sy Read error in ..
indicate possible file system trouble and should be referred to a UNIX system
administrator.
.Sh INTERFACE STABILITY
.Sy Committed
.Sh CODE SET INDEPENDENCE
.Sy Enabled
.Sh SEE ALSO
.Xr cd 1 ,
.Xr ksh 1 ,
.Xr ksh93 1 ,
.Xr sh 1 ,
.Xr shell_builtins 1 ,
.Xr attributes 7 ,
.Xr environ 7 ,
.Xr standards 7
.Sh NOTES
If you move the current directory or one above it,
.Nm
may not give the correct response.
Use the
.Xr cd 1
command with a full path name to correct this situation.


usr/src/cmd/pwd/pwd.c
/*
 * CDDL HEADER START
 *
 * The contents of this file are subject to the terms of the
 * Common Development and Distribution License, Version 1.0 only
 * (the "License").  You may not use this file except in compliance
 * with the License.
 *
 * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
 * or http://www.opensolaris.org/os/licensing.
 * See the License for the specific language governing permissions
 * and limitations under the License.
 *
 * When distributing Covered Code, include this CDDL HEADER in each
 * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
 * If applicable, add the following below this CDDL HEADER, with the
 * fields enclosed by brackets "[]" replaced with your own identifying
 * information: Portions Copyright [yyyy] [name of copyright owner]
 *
 * CDDL HEADER END
 */
/*
 *	Copyright 2005 Sun Microsystems, Inc.  All rights reserved.
 *	Use is subject to license terms.
 */

/*	Copyright (c) 1984, 1986, 1987, 1988, 1989 AT&T	*/
/*	  All Rights Reserved	*/

/*
 * Copyright 2026 Oxide Computer Company
 */

/*
 * Print working (current) directory
 *
 * POSIX eventually added two flags in issue 6, which are for -P and -L. -P is
 * basically the mode that has traditionally been used, which should resolve
 * symlinks. -L is a logical mode that is designed to interact with the shell.
 * Basically it says if $PWD is set and meets the following conditions it should
 * be used:
 *
 * 1. It is an absolute path.
 * 2. It does not have any '.' or '..' path components in it.
 * 3. It actually is the current working directory by comparing a stat of that
 *    and '.'.
 *
 * If any of these is not true then it is supposed to fall back to the
 * traditional behavior here.
 *
 * Finally there is one last wrinkle, the default behavior. POSIX mandates that
 * the equivalent of -L be done by default. However, both GNU and *BSD generally
 * follow the traditional default of behaving as though -P is the default. Given
 * the 40+ year history of this command, we maintain that default and do the
 * same thing as others do.
 */

#include	<stdio.h>
#include	<unistd.h>
#include	<stdlib.h>
#include	<limits.h>
#include	<locale.h>
#include	<string.h>
#include	<stdbool.h>
#include	<err.h>
#include	<stdnoreturn.h>
#include	<sys/stat.h>

static char name[PATH_MAX+1];

static noreturn void
usage(const char *fmt, ...)
{
	if (fmt != NULL) {
		va_list ap;

		va_start(ap, fmt);
		vwarnx(gettext(fmt), ap);
		va_end(ap);
	}

	(void) fprintf(stderr, gettext("Usage:  pwd [-L | -P]\n"));
	exit(1);
}

static const char *
getcwd_log(void)
{
	const char *cwd = getenv("PWD");
	if (cwd == NULL || cwd[0] != '/')
		return (NULL);

	/*
	 * Look for path components that aren't solely '.' and '..'.
	 */
	for (size_t i = 0; cwd[i] != '\0'; i++) {
		if (cwd[i] != '/' || cwd[i + 1] != '.')
			continue;

		if (cwd[i + 2] == '\0' || cwd[i + 2] == '/') {
			return (NULL);
		}

		if (cwd[i + 2] == '.' && (cwd[i + 3] == '/' ||
		    cwd[i + 3] == '\0')) {
			return (NULL);
		}
	}

	struct stat log, phys;
	if (stat(cwd, &log) != 0 || stat(".", &phys) != 0 ||
	    log.st_dev != phys.st_dev || log.st_ino != phys.st_ino) {
		return (NULL);
	}

	return (cwd);
}

int
main(int argc, char *argv[])
{
	int c;
	bool Lflag = false;
	const char *cwd = NULL;

	(void) setlocale(LC_ALL, "");
#if !defined(TEXT_DOMAIN)	/* Should be defined by cc -D */
#define	TEXT_DOMAIN "SYS_TEST"	/* Use this only if it weren't */
#endif
	(void) textdomain(TEXT_DOMAIN);

	while ((c = getopt(argc, argv, ":LP")) != -1) {
		switch (c) {
		case 'L':
			Lflag = true;
			break;
		case 'P':
			Lflag = false;
			break;
		case ':':
			usage("option -%c requires an argument", optopt);
		case '?':
			usage("unknown option -%c", optopt);
		}
	}

	/*
	 * Historically our implementation of pwd has never checked for this. To
	 * reduce the likelihood of this being a fatal error if someone somehow
	 * gets /usr/bin/pwd, we instead make this a warning.
	 */
	if (argc - optind > 0) {
		warnx(gettext("ignoring unexpected operands starting with %s"),
		    argv[optind]);
	}

	if (Lflag) {
		cwd = getcwd_log();
	}

	if (cwd == NULL) {
		cwd = getcwd(name, PATH_MAX + 1);
		if (cwd == NULL) {
			err(2, gettext("cannot determine current directory"));
		}
	}

	if (puts(cwd) == EOF) {
		err(EXIT_FAILURE, gettext("failed to write out current working "
		    "directory"));
	}

	exit(0);
}
